TS migration + Refactoring

Hofled opened the pull request:
# TS Migration + Refactoring

Added Typescript to the project and all necessary infrastructural stuff.
Tests are currently broken but they are WIP.
Everything else should function normally, while a lot of bugs were resolved.
Some of the code was migrated to Typescript already, while the majority was kept intact with minor changes (or bug fixes) to allow the migration.

## Things to look out for in this PR:
While going over the code, not a lot of logic has been re-written, but most of the stuff was moved \ renamed, and some bugs were fixed.

Try to focus on the structural changes and bug fixes, and not on the many file removals (as files were converted from `.js` to `.ts`.

A big area to focus on is the `models` folder, as the entire `Sequelize` setup has been converted into Typescript, but the database configuration and layout has been kept the same.
The model-specific logic functions have been nested under the corresponding modules, and are globally exported through `index.ts` as `[ModelName]Module` to be consumed through a reference to the `models/` directory (examples to such occurrences can be found in command files).

Hofled committed:
Added typescript to the project.

Hofled committed:
Replaced markdown with .gitkeep file.

Hofled committed:
Moved all code into the `src` directory.
Added typescript to the project.
Refactored sequelize models to use typescript.
Fixed bug in commands `index.js` file importing.

Hofled committed:
Added jest configuration for ts compatibility.

Hofled committed:
Fixed typo in table name.

Hofled committed:
added sequelizesrc configuration file.
added new `db` folder with db related files (migrations & seeders).
Added virtual function to fisherman.
Updated fisherman logic functions.
Added missing `findOrCreateByDiscordId` logic function to `User` model.
Exported new logic functions from db models.
Added sequelize-cli migration script to `package.json`.

Hofled committed:
Updated `README.md` with local db setup command.

Hofled committed:
Refactored db models to be accessed through typescript modules.
Refactored code and fixed bugs.
Refactored DB models usage in commands (now accesses through modules).
Added virtual mixins to the User model.
Removed redundant code from cheems.js

Hofled committed:
Added postinstall script to automatically run db migration after installation.

Hofled committed:
Updated documentation in README.md

Hofled committed:
Updated ReADME.md

Hofled committed:
Implemented enabled commands filtering using .env

Hofled committed:
Fixed bug where enabled commands parameter splitting was not done correctly.

Hofled committed:
Added .env documentation to support the latest command enabling feature.

Hofled committed:
Removed redundant comment.

Hofled committed:
Fixed bug where sell message was inappropriate if the inventory was worth 0.

Hofled committed:
Updated jest config file.
Updated README.md
Fixed bugs in Dice.js
Small refactor in cloneEnv.js to use `path` lib.

Hofled committed:
Converted clone test to typescript.
Partially migrated dice test to typescript (Constants class is still broken)
Removed test typescript subproject from root `tsconfig.json`
Partially migrated commandThrottlingService test to typescript.

Hofled committed:
temporarily updated gamble test load paths (needs to be converted to jest + typescript)

Hofled commented:
### Let's merge this bad boy in @VuongRon 
![4ujz4u](https://user-images.githubusercontent.com/9660580/105213609-95e49500-5b57-11eb-8e11-fe1acf3d4c38.jpg)

VuongRon commented on test/dice.test.ts:
```diff
@@ -108,7 +107,7 @@ describe ("Dice", () => {
 
     it ("should __NOT__ detect 'a Three of a kind' before 'Full House' in [2, 5, 2, 5, 2]", () => {
-      // Full House takes precedence ofer Three of a Kind, so we have to
+      // Full House takes precedence offer Three of a Kind, so we have to
```
👌 💯 🔥

DarkStoorM replied:
Should be `over` 🏃 💨

Hofled replied:
This is the highlight of the PR

Hofled replied:
resolved by commit ff1e8acb518c6081ddef5303fd6ec02575d43194

Hofled committed:
Fixed comment typo.

Hofled committed:
Fixed bug in resource command as an async functon was invoked without `await` before it.
Fixed imports to new model modules in resource.js.

Hofled committed:
Fixed imports in command files.
Added type to method definition in `media` model.

DarkStoorM committed:
Remove redundant Sequelize object initialization

DarkStoorM committed:
Bump Axios

DarkStoorM committed:
Add testing environment configuration

DarkStoorM committed:
Remove old services and tests

Channel Binding will be replaced with a simplified service
Command Throttling needs to be redesigned - unnecessary at the moment

DarkStoorM committed:
Use updated User Model

DarkStoorM committed:
Add a new Command type

DarkStoorM committed:
Change args type

Additional arguments always come as an array of strings - a result of message content split (space-separated command parameters)

DarkStoorM committed:
Set a reminder

Why is the TODO here?

DarkStoorM committed:
Convert command exporter to TS

DarkStoorM committed:
Add CommandCollector type

DarkStoorM committed:
Move Command type declaration outside to its own file

DarkStoorM committed:
Reorder the imports

DarkStoorM committed:
Use new CommandCollector type

DarkStoorM committed:
Use extracted CommandEnablingService

DarkStoorM committed:
Initialize channel bindings

DarkStoorM committed:
Add ChannelBinding / CommandEnabling services

DarkStoorM committed:
Remove old documentation and services

This will now use a new ChannelBinding service

DarkStoorM committed:
Update typings

DarkStoorM committed:
Document the initial abort, reformat the console logger

DarkStoorM committed:
Remove import redundancy

DarkStoorM committed:
Update the CommandCollection type naming

DarkStoorM committed:
Move bot commands to the client constructor

* adds TODO - reminder to rewrite the command importer

DarkStoorM committed:
Extract ExtendedClient class, remove unused import

DarkStoorM committed:
Update services description

adds a bit more descriptive comments

DarkStoorM committed:
Move ExtendedClient to /types

This should be a proper place :runner: :dash:

DarkStoorM committed:
Update CommandCollection import

removes unnecessary directory change while importing

Hofled committed:
Renamed db configuratoin json and updated reference.

Hofled committed:
Removed default value for `commands` property and updated documentation.

Hofled committed:
Refactored commands structure.
Added command factory for generating a command collection.
Updated command class with command appropriate logic.
Updated channel binding service.
Updated command enabling service to only return the set of enabled commands.

Hofled committed:
Removed usage of generics in `Command` type.

Hofled committed:
Fixed bug where `diceRolls` array was not initialized with 5 empty spots, as required.

Hofled committed:
Refactored usage of constants in the application.
Added command specific configuration jsons which can be imported and consumed by the command's logic.
Refactored `8ball` to typescript.
Extracted hardcoded responses into `eightBall.json`.
Removed previous `Constants` class and implementation.
Removed `config` from the `Command` class and `execute` method.

Hofled committed:
Added missing types and imports.

Hofled committed:
Updated readme with new description for creating commands.

Hofled committed:
Merge pull request #1 from Hofled/ts-darkstoorm

migrated additional TS refactors and infrastractural refactors

Hofled committed:
Added typing to the embedService.

dcepil reviewed (approved):
Looks good to me - if some things are broken we can fix them after merging this.

Hofled committed:
Updated README.md

Hofled committed:
Fixed typo and updated README.md

VuongRon reviewed (approved):
👴
